### DISCUSSION TITLE
How to use ratatui with an "emulated PTY"? Eg serving over an SSH server?

### DISCUSSION THREAD

**Discussion by samuela** [ratatui/ratatui#643] [Discussion]
Hi folks! 👋 I'm curious if ratatui support rendering to an "emulated PTY", eg. a PTY handled by an SSH server. I'm working on building a custom SSH server with [russh](https://github.com/warp-tech/russh/) that serves a TUI application over the connection. In an ideal world I'd like to be able to use hook everything together in-process such that ratatui renders to an SSH client's PTY.

Can ratatui handle such a usecase? What would be the best way to go about it?

**Comment by joshka** [comment]
This is probably something that would be implemented upstream in Crossterm, Termion (both of which support constructing a backend passing in anything that implements the `std::io::Write` trait), or termwiz (which works similarly under the hood).

There is example code for termion in the russh repo https://github.com/search?q=repo%3Awarp-tech%2Frussh+termion&type=code, but it looks like it's just the client side not the server side. It might be worth asking over on russh about Crossterm / Termion on the server side of things.

Alternatively if that approach doesn't work out you could write a backend that implements [`ratatui::backend:: Backend`](https://docs.rs/ratatui/latest/ratatui/backend/trait.Backend.html) over the rushh stuff. That would be pretty neat but more work perhaps. Taking a quick look at russh it looks like everything is async, which may introduce an extra level of difficulty if you're not adept in the ways of `Future`s and `Stream`s.

I look forward to seeing what you come up with.

**Reply by Eugeny** [reply]
~Heya, russh maintainer here. Having a pluggable backend is the ideal solution here, but something is amiss in `crossterm`'s implementation since it only seems to accept a pluggable `Write` output - but what about `stdin`/`SIGWINCH`/`termios`? All these are tunnelled over the SSH protocol.~

~`ratatui::Backend` is definitely doable but requires basically rewriting `crossterm`'s Linux implementation yourself~

Just as I was writing this I realized that `ratatui` doesn't provide input handling, hence there's nothing to implement here. 🤦  
@samuela you can basically just use a `CrosstermBackend::new(channel.into_stream())` for the output part. I don't know what crate you're using to handle user input, but you should look into whether that crate supports receiving events programmatically instead from the stdin's PTY.

**Reply by joshka** [reply]
Yeah, we just do the buffer -> backend part, the app handles the user input / app loop part.

**Comment by alexandrebrg** [comment]
Quick update on this discussion as russh maintainers recently created an example app with ratatui [here](https://github.com/warp-tech/russh/pull/249) :)

**Reply by ricott1** [reply]
Hi, I am the author of the linked example. I have an example app using this approach here: https://github.com/ricott1/sshattrick

**Reply by joshka** [reply]
I also just created a simple example of a Pong game
SSHattrick is pretty awesome btw.

**Reply by antoineco** [reply]
Another terminal game using `ratatui` and `russh` is [kartoffels](https://codeberg.org/pwy/kartoffels) :potato:
